policy: Ephemeral anchors

instagibbs opened the pull request:
TODO:

I'll be updating to segwit variant soon, with updated BIP text.

-----

Builds on top of https://github.com/bitcoin/bitcoin/pull/25038 for consideration of inclusion to the proposal. Requires V3, for simplicity of reasoning and usage. Implementation of idea written out at https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-October/021036.html

BIP text here: https://github.com/instagibbs/bips/blob/ephemeral_anchor/bip-ephemeralanchors.mediawiki

glozow committed:
[refactor] make some members MemPoolAccept-wide

No change in behavior.

For single transaction acceptance, this is a simple refactor:
Workspace::m_all_conflicting         -> MemPoolAccept::m_all_conflicts
Workspace::m_replacement_transaction -> MemPoolAccept::m_rbf
Workspace::m_conflicting_fees        -> MemPoolAccept::m_conflicting_fees
Workspace::m_conflicting_size        -> MemPoolAccept::m_conflicting_size
Workspace::m_replaced_transactions   -> MemPoolAccept::m_replaced_transactions

And local variables m_total_vsize and m_total_modified_fees are now
MemPoolAccept members so they can be accessed from PackageMempoolChecks.

We want these to be package-wide variables because
- Transactions could conflict with the same tx (just not the same
prevout), or their conflicts could share descendants.
- We want to compare conflicts with the package fee rather than
individual transaction fee.

glozow committed:
[packages/policy] implement package RBF

glozow committed:
[unit test] for CheckAncestorScores

glozow committed:
[doc] nVersion 3 and package rbf policies

glozow committed:
[policy] policy rules for nVersion=3

glozow committed:
[policy] allow V3 transactions under certain conditions

glozow committed:
[policy] check whether replacement is more incentive compatible

Avoid accepting replacements that are not more incentive compatible to
mine.  For now, as a conservative estimate, require that the minimum
between the transaction's individual feerate and ancestor feerate is
higher than the individual feerates of directly conflicting transactions
and the ancestor feerates of all original transactions.

Note that a package/transaction's ancestor feerate is not perfectly
representative of its incentive compatibility; it may overestimate (some
subset of the ancestors could be included by itself if it has other
high-feerate descendants or are themselves higher feerate than this
package/transaction). This is a conservative estimate and works for now.

Co-authored-by: Suhas Daftuar <sdaftuar@chaincode.com>

glozow committed:
[functional test] v3 transaction submission

glozow committed:
[test framework] parameterize version in create_self_transfer

instagibbs committed:
Add OP_TRUE as a standard output type

This will be used in following commits
as the flag for ephemeral anchors, allowing
dust values, but also requiring
the output to be spend in the same relay package.

glozow committed:
[refactor] rename entries to be more descriptive

glozow committed:
[doc] cpfp carveout is excluded in packages

The behavior is not new, but this rule exits earlier than before.
Previously, a carve out could have been granted in PreChecks() but then
nullified in PackageMempoolChecks() when CheckPackageLimits() is called
with the default limits.

luke-jr commented:
Concept NACK as a mere policy. Without consensus enforcement, cost is a full UTXO which may potentially never be spent. `OP_TRUE` does not have a weight that reflects this high cost.

instagibbs commented:
@luke-jr Can you elaborate more on what the potential issues are?

This policy defines that any transaction with such an output must be 0-fee, which means there's little incentive to mine it on its own, but only as part of a CPFP spend of that output.

Assuming the PR is correct and being run on the network, there should be no way(aside from miners prioritizing the parent explicitly) to leaving the utxo in the uxto set. Prioritizing the parent could also be disallowed easily.

luke-jr commented:
The main issue I see is that it has a very low weight, but rather high cost. There is no consensus rule requiring it to be spent immediately.

Couldn't it be CPFP'd by spending another of its outputs?

instagibbs commented:
The argument here is that dust limits are already policy only, and this is basically a "cut-through" of  potentially dust outputs. It's no practical change of dust policy, except in the case where a miner prioritizes a parent tx with this kind of output after its included in mempool(it's in the test cases)

JeremyRubin commented:
@luke-jr  either you do a consensus fork and everything that entails for something like tx sponsors, or your enable in policy only something like ephemeral anchors.

One option that could be taken would be to make the OP_TRUE script contain an upgradable NOP that we document as reserved to in the future treat as a pre-execution parsed flag with a consensus must-spend rule, and then we'd be able to (if there was wide adoption of ephemeral anchors and we saw utxo bloat) add a consensus rule to enforce them as must-spend.

luke-jr commented:
>One option that could be taken would be to make the OP_TRUE script contain an upgradable NOP that we document as reserved to in the future treat as a pre-execution parsed flag with a consensus must-spend rule, and then we'd be able to (if there was wide adoption of ephemeral anchors and we saw utxo bloat) add a consensus rule to enforce them as must-spend.

I think this would be better, since it allows a real OP_TRUE to be given a more cost-based weight in a future softfork.

MarcoFalke commented:
It would also be possible to deploy a consensus change to treat them as unspendable, unless they are spent in the block they are created in, but it seems odd to enforce policy with consensus, and this is likely the wrong thread to discuss anyway.

JeremyRubin commented:
@MarcoFalke that is an inverse timelock, which has nasty consequences for reorg safety and dependent transactions.

instagibbs commented:
@JeremyRubin including a NOP would just be to not "camp" a raw single byte value push for a future soft fork? Making sure I didn't miss the underlying motivation.

JeremyRubin commented:
more or less, you might also be able to do a 0 byte segwit v1 script or something too.

Otherwise you end up with some "specific script template matching upgrade", which is IMO messy & risks being uneccessarily confiscatory.

ajtowns commented:
Camping an OP_NOP and preventing it from being used for any other future soft fork seems worse than camping OP_TRUE.

If in the future we have a soft fork that makes small scriptPubKeys more costly to limit utxo bloat with an exemption for ephmeral outputs that are spent in the same block, which should just apply both rules to all scriptPubKeys, whether they're OP_TRUE, OP_2, or OP_NOP5 OP_TRUE...

glozow committed:
[test util] CreateValidMempoolTransaction version parameter, always signal bip125

glozow committed:
[doc] doc assumptions package RBF

Try to prevent accidental future breakage and document some assumptions
that would break package RBF if changed.

glozow committed:
[validation] refactor sub-package submissions, call AcceptSingle when only 1 tx

Avoid calling PackageMempoolChecks() when there is only 1 transaction.
Note to reviewers: there is a slight change in the error type returned,
as shown in the txpackage_tests change.  When a transaction is the last
one left in the package and its fee is too low, this returns a PCKG_TX
instead of PCKG_POLICY. This interface is clearer;
"package-fee-too-low" for 1 transaction would be a bit misleading.

instagibbs commented:
I'll rebase on top of https://github.com/bitcoin/bitcoin/pull/25038 once it leaves draft mode, since I have to update my tests to new V3 packages.

naumenkogs commented:
What will happen at the p2p level, if the attacker manages to submit a transaction with ephemeral outputs to many nodes in the network without a child?

It looks like its wtxid will make it into `m_recent_rejects`, and then the network won't reconsider the tx even when the child is attached?

instagibbs commented:
@naumenkogs Ideally I would hope it would be treated like any other too-low-fee transaction, and be re-evaluated again in the package context. I'll double check, thanks!

instagibbs commented on test/functional/mempool_ephemeral_anchor.py:
```diff
@@ -0,0 +1,341 @@
+        self.assert_mempool_contents(expected=package_txns3, unexpected=[])
+
+        # All ephemeral anchors this time, make sure old child evicted
+        package_hex5, package_txns5 = self.create_simple_package(parent_coins=parent_coins, parent_fee=0, child_fee=DEFAULT_FEE*2, version=3)
```
todo: double-check we can't get blinded to a parent tx if submitted separately

ajtowns commented on test/functional/mempool_ephemeral_anchor.py:
```diff
@@ -331,11 +331,12 @@ def test_xor_rbf(self):
         self.assert_mempool_contents(expected=[second_package_txns[1]], unexpected=[])
 
-        # Parent can still be prioritised
+        # Parent can't be prioritised; RPC call succeeds but is modified fees are ignored if it has EAs
```
"i***t***s modified fees" ?

instagibbs replied:
removed the test since it's duplicated by v3 work which now trims lower than min relay fee txns

ajtowns commented on src/txmempool.cpp:
```diff
@@ -402,6 +407,11 @@ void CTxMemPoolEntry::UpdateDescendantState(int64_t modifySize, CAmount modifyFe
 
 void CTxMemPoolEntry::UpdateAncestorState(int64_t modifySize, CAmount modifyFee, int64_t modifyCount, int64_t modifySigOps)
 {
+    // We don't want to encourage miners to mine these transacitons alone. Instead
```
"transac***ti***ons"

instagibbs replied:
no longer changing this

glozow committed:
[policy] allow individual v3 txns to be below min relay feerate

As long as they are otherwise paid for, i.e. through package CPFP.
If a v3 transaction loses its sponsor, we can evict them with no trouble
because it will not have other descendants or ancestors to make the
feerate assessment more complicated.

glozow committed:
[mempool] evict everything below min relay fee in TrimToSize()

At this point it's not expected that there are any such transactions,
except from reorgs and possibly when loading from mempool.dat.

glozow committed:
[test] package rbf

instagibbs committed:
Functional tests for ephemeral anchors

Does not have test coverage for making sure EA are spent,
and package RBF cases, neither of which exist currently.

instagibbs commented:
Rebased on top of latest https://github.com/bitcoin/bitcoin/pull/25038

1) simplified spends checking due to 2-txn topology for better error reporting
2) simplified tests, no longer messing with test utils at all

instagibbs commented:
alternative commit that makes it OP_TRUE instead: https://github.com/instagibbs/bitcoin/commit/bf7676da978188ebf759288a147f79eb41bf80cc

instagibbs committed:
Allow dust anchor outputs

Anchor outputs are not yet required
to be spent in same mempool package.

instagibbs commented:
Switched to OP_TRUE, as it was less confusing for at least a few people, and only requires light test editing. Had to rebase on top of master for trivial merge conflict.

instagibbs commented on src/policy/policy.cpp:
```diff
@@ -142,8 +142,12 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         } else if (IsDust(txout, dust_relay_fee)) {
-            reason = "dust";
-            return false;
+            // We ensure ephemeral output is spent in the same mempool package
```
suggested in other PR: "switch the ordering? else if (whichType != ANCHOR && IsDust())"

instagibbs committed:
Add -ephemeralanchors which when set to false disallows anchor outputs

instagibbs committed:
Add anchor argument to transaction creation rpcs

instagibbs committed:
Allow one ephemeral anchor per transaction

instagibbs committed:
Enforce that ephemeral anchor transactions are 0 base fee

instagibbs committed:
GetDustThreshold: Make anchor output dust threshold 0 explicitly

instagibbs committed:
Only allow ephemeral anchors in v3 transactions

instagibbs committed:
Require ephemeral anchors to be spent in mempool atomically

Either in the same relay package, or by transactions RBFing
the CPFP.

instagibbs committed:
Add test for more complicated ephemeral anchor scenarios

instagibbs commented:
I reworked this PR with changes incorporated from https://github.com/bitcoin-inquisition/bitcoin/pull/23

instagibbs commented:
Should add a test showing that node restart dumps ephemeral transactions since re-admission to the mempool is done single tx by single tx only.

instagibbs committed:
Add test for ephemeral anchors on node restart

instagibbs committed:
Test failure to evict ephemeral transaction on node restart

instagibbs commented:
https://github.com/bitcoin/bitcoin/pull/26403/commits/a1982f5773690bc0c74d1bb932cb7552b00aacaf shows that current re-submission logic on node restart fails to evict ephemeral transaction. 

Until there is package-aware re-submission, should these transactions just be dropped on the floor? These types of transactions are opting into fairly aggressive fee bumping patterns, maybe that's an ok first cut.

instagibbs commented:
note to self: after rebased onto BIP331, need to make sure rejected ephemeral anchor tx is put in `m_recent_rejects_reconsiderable` filter instead of `m_recent_rejects`

instagibbs commented on src/validation.cpp:
```diff
@@ -834,6 +834,11 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
 
+    // We check for 0-base-fee transaction here now that we have access to ws.m_base_fees.
+    if (HasPayToAnchor(tx) && ws.m_base_fees != 0) {
+        return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "invalid-ephemeral-fee");
```
needs to return a code like `TX_LOW_FEE` from bip331 post-rebase

instagibbs commented on src/policy/v3_policy.cpp:
```diff
@@ -126,13 +126,19 @@ std::optional<std::string> CheckEphemeralSpends(const CTransactionRef& ptx,
+    /* Only allow in package feerate context */
+    if (!package_context) {
+        /* Allows re-evaluation in package context */
+        return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "missing-ephemeral-spends");
```
should return code like `TX_LOW_FEE` re:BIP331 to avoid blinding relay of an otherwise valid parent

instagibbs commented:
Random thought:

OP_TRUE is txid-malleable by block maker. e.g., they could theoretically add in superfluous pushes in the `scriptSig` and this would not fail validation because `CLEANSTACK` is a policy-only rule.

In this particular V3 context, it just means your CPFP could change txid, wouldn't do anything else. Maybe in a more general context if we weren't restricted to V3 topology this would be an issue? In the context of keyless anchors, I'm not sure that will ever make sense anyways.

The "fix" would be: `OP_DEPTH OP_0 OP_EQUAL`, making it consensus-non-malleable at the cost of 8 additional WU. Seems like a bad tradeoff, noting anyways.

instagibbs committed:
Disallow individual submission of ephemeral transaction

instagibbs committed:
Test to make sure that EA txns are evicted

glozow commented:
> @instagibbs pushed 1 commit.
>    [faeed68](https://github.com/bitcoin/bitcoin/commit/faeed687e5cde5e32750d93818dd1d4add837f24) nice crash bro

hacked a quick-ish fix on top of this branch [here](https://github.com/glozow/bitcoin/tree/2023-08-10-fix-crash-bro). As discussed offline, the general problem is with calling `TrimToSize` in the middle of package validation (meaning something can go from already-in-mempool to evicted, or just-submitted to evicted) so a coin may be cached in `m_view` but then it disappears from the mempool/temporary coins in `m_viewmempool`. Crash is in `CheckInputsFromMempoolAndCache` which asserts the coin exists, but it doesn't anymore. Will open a separate PR to address this since it's also a problem pre-v3.

glozow committed:
[validation] make PackageMempoolAcceptResult members mutable

After the PackageMempoolAcceptResult is returned from
AcceptMultipleTransactions, leave room for results to change due to
LimitMempool() eviction.

glozow committed:
[refactor] back-filling m_tx_results in AcceptPackage

glozow committed:
[validation] don't LimitMempoolSize in any subpackage submissions

instagibbs committed:
Return complete reason for package submission failure

instagibbs committed:
Make sure test_sponsor_swap is testing the desired behavior

whitslack commented:
What's the point of having a non-malleable version? Okay, so a miner cannot add extra pushes to the input script that spends the anchor, but there is nothing to stop the miner from mining an alternative child transaction that spends *only* the anchor. Thus, the party that had broadcast the original child transaction will be forced to construct a new spending transaction with a different TxID anyway. Indeed, anyone (not only a miner) is free to propose a replacement child that spends *only* the anchor, so it will never be okay to depend upon the TxID of a transaction that spends an ephemeral anchor. So again I ask, what is the point of having a non-malleable ephemeral anchor?

instagibbs commented:
> Thus, the party that had broadcast the original child transaction will be forced to construct a new spending transaction with a different TxID anyway

Canonical example here which motivated the change was that of LN channel splicing being used to cpfp bump another transaction, would could be another channel's commitment transaction itself. This requires the ephemeral anchor-spending tx(the splice) to retain txid stability otherwise funds are locked up in the new funding output.

I understand that not all use-cases require such things, but two points:
1) It's easy to expand standardness, and really hard to restrict. In deploying one version, we'll likely learn lessons we'd want to apply to other versions. If we see tons of update from wallets that explicitly don't care about txid stability? That's a signal as well.
2) "I don't see the problem" is much weaker than "there is no problem" and I'd rather be able to reason via modern bitcoin rules

whitslack commented:
> This requires the ephemeral anchor-spending tx(the splice) to retain txid stability otherwise funds are locked up in the new funding output.

The point I was missing is that a non-malleable anchor ensures that the *other* output(s) do **not** get spent in a transaction whose TxID was not known in advance of signing. Thank you for the answer.
